### PR TITLE
Function provider for simple methods registration as C1 functions

### PR DESCRIPTION
**Pull request by vadym-hyryn**
Hey:) This is the integration of ServiceProvider from EM to C1 with updates
The name had to be changed because not only SP used now, but still it can be updated somehow

### CODE REVIEW

File: Composite/Data/DataMetaDataFacade.cs [hunk 1]
```diff
@@ -198,11 +198,11 @@ public static DataTypeDescriptor GetDataTypeDescriptor(Guid dataTypeId, bool all
...
-                if (!AssemblyFacade.AssemblyPotentiallyUsesType(assembly, typeof(IData)))
-                {
-                    // Ignoring assemblies that aren't referencing Composite.dll
-                    continue;
-                }
+                //if (!AssemblyFacade.AssemblyPotentiallyUsesType(assembly, typeof(IData)))
+                //{
+                //    // Ignoring assemblies that aren't referencing Composite.dll
+                //    continue;
+                //}
```
Comment (napernik) [inline Composite/Data/DataMetaDataFacade.cs:206]: This there a reason for this code block to be commented out?

File: Composite/Plugins/Functions/FunctionProviders/CodeBasedFunctionProvider/CodeBasedFunction.cs [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+            }
+            if (MethodInfo.IsStatic)
+            {
+                return MethodInfo.Invoke(null, arguments.ToArray());
+            }
+
+            var serviceObject = ServiceLocator.GetService(Type);
+            return serviceObject == null
+                ? MethodInfo.Invoke(Activator.CreateInstance(Type), arguments.ToArray())
+                : MethodInfo.Invoke(serviceObject, arguments.ToArray());
```
Comment (napernik) [inline Composite/Plugins/Functions/FunctionProviders/CodeBasedFunctionProvider/CodeBasedFunction.cs:52]: Can be simplified:

```
object instance = MethodInfo.IsStatic ? null : (ServiceLocator.GetService(Type) ?? Activator.CreateInstance(Type));
MethodInfo.Invoke(instance, arguments.ToArray())
```

File: Composite/Plugins/Functions/FunctionProviders/CodeBasedFunctionProvider/CodeBasedFunctionEntityToken.cs [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+using Composite.Functions;
+
+namespace Composite.Plugins.Functions.FunctionProviders.CodeBasedFunctionProvider
+{
+    [SecurityAncestorProvider(typeof(StandardFunctionSecurityAncestorProvider))]
+    internal class CodeBasedFunctionEntityToken : EntityToken
+    {
+        internal CodeBasedFunctionEntityToken(CodeBasedFunction function)
+        {
+            Id = string.Concat(function.Namespace, ".", function.Name);
```
Comment (napernik) [inline Composite/Plugins/Functions/FunctionProviders/CodeBasedFunctionProvider/CodeBasedFunctionEntityToken.cs:11]: To mach the existing style $"{function.Namespace}.{function.Name}"

File: Composite/Data/DataMetaDataFacade.cs (hunk 1 above)
Comment (vadym-hyryn) [reply Composite/Data/DataMetaDataFacade.cs]: Agree, revert this change

File: Composite/Plugins/Functions/FunctionProviders/CodeBasedFunctionProvider/CodeBasedFunction.cs (hunk 2 above)
Comment (vadym-hyryn) [reply Composite/Plugins/Functions/FunctionProviders/CodeBasedFunctionProvider/CodeBasedFunction.cs]: Thanks, applied

File: Composite/Plugins/Functions/FunctionProviders/CodeBasedFunctionProvider/CodeBasedFunctionEntityToken.cs (hunk 3 above)
Comment (vadym-hyryn) [reply Composite/Plugins/Functions/FunctionProviders/CodeBasedFunctionProvider/CodeBasedFunctionEntityToken.cs]: Updated

**Pull request merged**